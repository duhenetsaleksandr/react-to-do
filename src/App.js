import './styles/wrapper.scss';
import React, { useState, useEffect } from 'react';
import axios from 'axios';
import Header from "./components/Header";
import Main from "./components/Main";
import Footer from "./components/Footer";
import Loader from './components/Loader';
import * as constants from './constants/constants';
import fetchTodo from './helpers';

export default function App () {
    const [ loader, setLoader ] = useState(false);
    const [ todos, setTodos ] = useState([]);

    const showLoader = () => setLoader(true);
    const hideLoader = () => setLoader(false);

    useEffect(() => {
        (async () => {
            try {
                showLoader();
                const data = await fetchTodo();
                setTodos(data);
            } catch (error) {
                console.error(error);
            } finally {
                hideLoader();
            }
        })();
    }, []);

    const editTodoInState = (data) => {
        setTodos((todos) => todos.map((todo) => todo.id === data.id ? data : todo));
    }

    const toggleTodo = async (id) => {
        try {
            showLoader();
            const todo = todos.find((todo) => todo.id === id);
            const body = { ...todo, completed: !todo.completed };
            const response = await axios.put(`${constants.API_URL}/${id}`, body);
            editTodoInState(response.data);
        } finally {
            hideLoader();
        }
    }

    const deleteTodo = async (id) => {
        try {
            showLoader();
            const response = await axios.delete(`${constants.API_URL}/${id}`);
            if (response.status === 200) {
                setTodos((todos) => todos.filter((todo) => todo.id !== id));
            }
        } finally {
            hideLoader();
        }
    }

    const createTodo = async (title) => {
        try {
            showLoader();
            const body = { title, completed: false };
            const response = await axios.post(constants.API_URL, body);
            setTodos((todos) => [ ...todos, response.data ]);
        } finally {
            hideLoader();
        }
    }

    const editTodo = async (id, title) => {
        try {
            showLoader();
            const todo = todos.find((todo) => todo.id === id);
            const body = {...todo, title};
            const response = await axios.put(`${constants.API_URL}/${todo.id}`, body);
            editTodoInState(response.data);
        } finally {
            hideLoader();
        }
    }

    return (
        <div className="wrapper">
            { loader && <Loader/> }
            <Header/>
            <Main
                todos={todos}
                toggleTodo={toggleTodo}
                deleteTodo={deleteTodo}
                createTodo={createTodo}
                editTodo={editTodo}
            />
            <Footer/>
        </div>
    );
}
