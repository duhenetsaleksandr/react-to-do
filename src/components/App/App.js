import React, { useState, useEffect } from 'react';
import { IntlProvider } from 'react-intl';
import './App.scss';
import ThemeContext from 'Context/ThemeContext';
import Header from "components/Header";
import Main from "components/Main";
import Footer from "components/Footer";
import Loader from 'components/common/Loader';
import { fetchTodoAPI, createTodoAPI, editTodoAPI, deleteTodoAPI } from 'helpers/requestAPI';
import { LOCALES } from 'i18n/locales';
import { messages } from 'i18n/messages';

const initialState = {
    isLoading: false,
    todos: [],
    darkTheme: true,
    locale: LOCALES.ENGLISH,
};

const App = () => {
    const [ state, setState ] = useState(initialState);

    useEffect(() => {
        initialTodos();
    }, []);

    const initialTodos = async () => {
        try {
            setState((prevState) => ({ ...prevState, isLoading: true }));
            const data = await fetchTodoAPI();
            setState((prevState) => ({ ...prevState, todos: data, isLoading: false }));
        } catch {
            setState((prevState) => ({ ...prevState, isLoading: false }));
        }
    };

    const editTodoInState = (data) => {
        setState((prevState) => (
            { ...prevState, todos: prevState.todos.map(todo => todo.id === data.id ? data : todo), isLoading: false }
        ));
    }

    const toggleTodo = async (id) => {
        try {
            setState((prevState) => ({ ...prevState, isLoading: true }));
            const todo = state.todos.find((todo) => todo.id === id);
            const body = { ...todo, completed: !todo.completed };
            const data = await editTodoAPI(body);
            editTodoInState(data);
        } catch {
            setState((prevState) => ({ ...prevState, isLoading: false }));
        }
    }

    const deleteTodo = async (id) => {
        try {
            setState((prevState) => ({ ...prevState, isLoading: true }));
            const status = await deleteTodoAPI(id);
            if (status === 200) {
                setState((prevState) => (
                    { ...prevState, todos: prevState.todos.filter((todo) => todo.id !== id), isLoading: false }
                ));
            }
        } catch {
            setState((prevState) => ({ ...prevState, isLoading: false }));
        }
    }

    const createTodo = async (title) => {
        try {
            setState((prevState) => ({ ...prevState, isLoading: true }));
            const data = await createTodoAPI(title);
            setState((prevState) => (
                { ...prevState, todos: [ ...prevState.todos, data ], isLoading: false }
            ));
        } catch {
            setState((prevState) => ({ ...prevState, isLoading: false }));
        }
    }

    const editTodo = async (id, title) => {
        try {
            setState((prevState) => ({ ...prevState, isLoading: true }));
            const todo = state.todos.find((todo) => todo.id === id);
            const body = {...todo, title};
            const data = await editTodoAPI(body);
            editTodoInState(data);
        } catch {
            setState((prevState) => ({ ...prevState, isLoading: false }));
        }
    }

    const switchTheme = () => {
        setState((prevState) => ({ ...prevState, darkTheme: !prevState.darkTheme }));
    };

    const switchLang = ({ target: { value } }) => {
        setState((prevState) => ({ ...prevState, locale: value }));
    };

    return (
        <ThemeContext.Provider value={{ darkTheme: state.darkTheme, switchTheme }}>
            <IntlProvider messages={messages[state.locale]} locale={state.locale} defaultLocale={LOCALES.ENGLISH}>
                <div className="wrapper">
                    { state.isLoading && <Loader/> }
                    <Header currentLang={state.locale} onChangeLocale={switchLang}/>
                    <Main
                        todos={state.todos}
                        toggleTodo={toggleTodo}
                        deleteTodo={deleteTodo}
                        createTodo={createTodo}
                        editTodo={editTodo}
                    />
                    <Footer/>
                </div>
            </IntlProvider>
        </ThemeContext.Provider>
    );
}

export default App;
