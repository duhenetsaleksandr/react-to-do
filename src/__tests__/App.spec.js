import React from 'react';
import axios from 'axios';
import { act } from "@testing-library/react";
import App from '../App';

jest.mock('axios');

const todos = [
    {
        userId: 1,
        id: 1,
        title: 'delectus aut autem',
        completed: false
    },
    {
        userId: 1,
        id: 2,
        title: 'quis ut nam facilis et officia qui',
        completed: false
    },
    {
        userId: 1,
        id: 3,
        title: 'fugiat veniam minus',
        completed: false
    },
];

const newTodo = {
    id: 201,
    completed: false,
    title: 'new create todo',
};

describe('should render App component', () => {
    describe('should success fetch data from API', () => {
        let component;
        beforeEach(async () => {
            axios.get.mockImplementation(() => Promise.resolve({ data: todos }));

            await act(async () => {
                component = mount(<App/>);
            });
            component.update();
        });

        it('should success fetch data after mounted App component and set data in state', async () => {
            expect(axios.get).toHaveBeenCalledTimes(1);
            expect(component.find('Main').getElement().props.todos).toEqual(todos);
        });

        it('should success toggle completed status', async () => {
            axios.put.mockImplementation((url, body) => Promise.resolve({ data: body }));

            await act(async () => {
                component.find('Main').getElement().props.toggleTodo(todos[0].id);
            });
            component.update();
            const changeTodos = [...todos];
            changeTodos[0].completed = !changeTodos[0].completed;
            expect(component.find('Main').getElement().props.todos).toEqual(changeTodos);
        });

        it('should fail toggle completed status', async () => {
            axios.put.mockImplementation(() => Promise.reject());

            await act(async () => {
                component.find('Main').getElement().props.toggleTodo(todos[0].id);
            });
            component.update();
            expect(component.find('Main').getElement().props.todos).toEqual(todos);
        });

        it('should success delete todo', async () => {
            axios.delete.mockImplementation(() => Promise.resolve({ status: 200 }));

            await act(async () => {
                component.find('Main').getElement().props.deleteTodo(todos[0].id);
            });
            component.update();
            expect(component.find('Main').getElement().props.todos).toHaveLength(todos.length - 1);
        });

        it('should fail delete todo', async () => {
            axios.delete.mockImplementation(() => Promise.reject());

            await act(async () => {
                component.find('Main').getElement().props.deleteTodo(todos[0].id);
            });
            component.update();
            expect(component.find('Main').getElement().props.todos).toHaveLength(todos.length);
        });

        it('should fail delete todo if response status not equal 200', async () => {
            axios.delete.mockImplementation(() => Promise.resolve({ status: 404 }));

            await act(async () => {
                component.find('Main').getElement().props.deleteTodo(todos[0].id);
            });
            component.update();
            expect(component.find('Main').getElement().props.todos).toHaveLength(todos.length);
        });

        it('should success edit todo', async () => {
            axios.put.mockImplementation((url, body) => Promise.resolve({ data: { ...body } }));

            await act(async () => {
                component.find('Main').getElement().props.editTodo(todos[0].id, todos[0].title + ' {{edited}}');
            });
            component.update();
            const changeTodos = [...todos];
            changeTodos[0].title = changeTodos[0].title + ' {{edited}}';
            expect(component.find('Main').getElement().props.todos).toEqual(changeTodos);
        });

        it('should fail edit todo', async () => {
            axios.put.mockImplementation(() => Promise.reject());

            await act(async () => {
                component.find('Main').getElement().props.editTodo(todos[0].id, todos[0].title + ' {{edited}}');
            });
            component.update();
            expect(component.find('Main').getElement().props.todos).toEqual(todos);
        });

        it('should success create new todo', async () => {
            axios.post.mockImplementation((url, body) => Promise.resolve({ data: { ...body, id: newTodo.id } }));

            await act(async () => {
                component.find('Main').getElement().props.createTodo(newTodo.title);
            });
            component.update();
            expect(component.find('Main').getElement().props.todos).toEqual([...todos, newTodo]);
        });


        it('should fail create new todo', async () => {
            axios.post.mockImplementation(() => Promise.reject());

            await act(async () => {
                component.find('Main').getElement().props.createTodo(newTodo.title);
            });
            component.update();
            expect(component.find('Main').getElement().props.todos).toEqual(todos);
        });
    });

    describe('should fail fetch data from API', () => {
        it('should fail fetch data after mounted App component and set empty array in state', async () => {
            axios.get.mockImplementation(() => Promise.reject());
            let component;
            await act(async () => {
                component = mount(<App/>);
            });
            component.update();
            expect(axios.get).toHaveBeenCalledTimes(1);
            expect(component.find('Main').getElement().props.todos).toEqual([]);
        });
    });

    describe('App component snapshot', () => {
        it('should render App component snapshot if fetch data from API success', async () => {
            axios.get.mockImplementation(() => Promise.resolve({ data: todos }));

            let component;
            await act(async () => {
                component = shallow(<App/>);
            });
            component.update();
            expect(component).toMatchSnapshot();
        });

        it('should render App component snapshot if fetch data from API fail', async () => {
            axios.get.mockImplementation(() => Promise.reject());

            let component;
            await act(async () => {
                component = shallow(<App/>);
            });
            component.update();
            expect(component).toMatchSnapshot();
        });
    });

    describe('On ThemeContext', () => {
        let component;

        beforeEach(async () => {
            await act(async () => {
                component = shallow(<App/>);
            });
            component.update();
        });

        it('should contain context provider', async () => {
            const provider = component.find('ContextProvider');
            expect(provider).toHaveLength(1);
        });

        it('should change theme', async () => {
            const provider = component.find('ContextProvider');
            const theme = provider.getElement().props.value.darkTheme;
            act(() => {
                provider.getElement().props.value.switchTheme();
            });
            component.update();
            expect(component.find('ContextProvider').getElement().props.value.darkTheme).toBe(!theme);
        });
    });
});
