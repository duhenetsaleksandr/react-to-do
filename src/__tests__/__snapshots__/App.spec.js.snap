// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render App component App component snapshot should render App component snapshot if fetch data from API fail 1`] = `
<ContextProvider
  value={
    Object {
      "darkTheme": true,
      "switchTheme": [Function],
    }
  }
>
  <IntlProvider
    defaultFormats={Object {}}
    defaultLocale="en-US"
    fallbackOnEmptyString={true}
    formats={Object {}}
    locale="en-US"
    messages={
      Object {
        "button_add": "Add task",
        "button_edit": "Edit task",
        "copyright": "©Copyright. All rights deserved.",
        "placeholder": "Enter new todo",
      }
    }
    onError={[Function]}
    textComponent={Symbol(react.fragment)}
  >
    <div
      className="wrapper"
    >
      <Header
        currentLang="en-US"
        onChangeLocale={[Function]}
      />
      <Main
        createTodo={[Function]}
        deleteTodo={[Function]}
        editTodo={[Function]}
        todos={Array []}
        toggleTodo={[Function]}
      />
      <Footer />
    </div>
  </IntlProvider>
</ContextProvider>
`;

exports[`should render App component App component snapshot should render App component snapshot if fetch data from API success 1`] = `
<ContextProvider
  value={
    Object {
      "darkTheme": true,
      "switchTheme": [Function],
    }
  }
>
  <IntlProvider
    defaultFormats={Object {}}
    defaultLocale="en-US"
    fallbackOnEmptyString={true}
    formats={Object {}}
    locale="en-US"
    messages={
      Object {
        "button_add": "Add task",
        "button_edit": "Edit task",
        "copyright": "©Copyright. All rights deserved.",
        "placeholder": "Enter new todo",
      }
    }
    onError={[Function]}
    textComponent={Symbol(react.fragment)}
  >
    <div
      className="wrapper"
    >
      <Header
        currentLang="en-US"
        onChangeLocale={[Function]}
      />
      <Main
        createTodo={[Function]}
        deleteTodo={[Function]}
        editTodo={[Function]}
        todos={Array []}
        toggleTodo={[Function]}
      />
      <Footer />
    </div>
  </IntlProvider>
</ContextProvider>
`;
